src/spawn.cpp

```diff
@@ -80,7 +80,78 @@ bool Spawns::loadFromXml(const std::string& filename)
 		Spawn& spawn = spawnList.front();
 
 		for (auto childNode : spawnNode.children()) {
-			if (strcasecmp(childNode.name(), "monster") == 0) {
+			if (strcasecmp(childNode.name(), "monsters") == 0) {
+				Position pos(
+					centerPos.x + pugi::cast<uint16_t>(childNode.attribute("x").value()),
+					centerPos.y + pugi::cast<uint16_t>(childNode.attribute("y").value()),
+					centerPos.z
+				);
+
+				int32_t interval = pugi::cast<int32_t>(childNode.attribute("spawntime").value()) * 1000;
+				if (interval < MINSPAWN_INTERVAL) {
+					std::cout << "[Warning - Spawns::loadFromXml] " << pos << " spawntime can not be less than " << MINSPAWN_INTERVAL / 1000 << " seconds." << std::endl;
+					continue;
+				} else if (interval > MAXSPAWN_INTERVAL) {
+					std::cout << "[Warning - Spawns::loadFromXml] " << pos << " spawntime can not be more than " << MAXSPAWN_INTERVAL / 1000 << " seconds." << std::endl;
+					continue;
+				}
+
+				size_t monstersCount = std::distance(childNode.children().begin(), childNode.children().end());
+				if (monstersCount == 0) {
+					std::cout << "[Warning - Spawns::loadFromXml] " << pos << " empty monsters set." << std::endl;
+					continue;
+				}
+
+				uint16_t totalChance = 0;
+				spawnBlock_t sb;
+				sb.pos = pos;
+				sb.direction = DIRECTION_NORTH;
+				sb.interval = interval;
+				sb.lastSpawn = 0;
+
+				for (auto monsterNode : childNode.children()) {
+					pugi::xml_attribute nameAttribute = monsterNode.attribute("name");
+					if (!nameAttribute) {
+						continue;
+					}
+
+					MonsterType* mType = g_monsters.getMonsterType(nameAttribute.as_string());
+					if (!mType) {
+						std::cout << "[Warning - Spawn::loadFromXml] " << pos << " can not find " << nameAttribute.as_string() << std::endl;
+						continue;
+					}
+
+					uint16_t chance = 100 / monstersCount;
+					pugi::xml_attribute chanceAttribute = monsterNode.attribute("chance");
+					if (chanceAttribute) {
+						chance = pugi::cast<uint16_t>(chanceAttribute.value());
+					}
+
+					if (chance + totalChance > 100) {
+						chance = 100 - totalChance;
+						totalChance = 100;
+						std::cout << "[Warning - Spawns::loadFromXml] " << mType->name << ' ' << pos << " total chance for set can not be higher than 100." << std::endl;
+					} else {
+						totalChance += chance;
+					}
+
+					sb.mTypes.push_back({mType, chance });
+				}
+
+				if (sb.mTypes.empty()) {
+					std::cout << "[Warning - Spawns::loadFromXml] " << pos << " empty monsters set." << std::endl;
```

Comment:
This is a little bit of a lie, since the set is not empty, but contains non existent monster(s) only 🤔